Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,346 @@
-        XCTAssertNil(result);
-        // We currently require every document read to also be written.
-        // TODO(b/34879758): Fix this check once we drop that requirement.
-        XCTAssertNotNil(error);
-        XCTAssertEqual(error.code, FIRFirestoreErrorCodeInvalidArgument);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence %@, to succeed, but got %ld.",
+                                       [self listStages], [error code]];
+        [self->_testCase assertNilErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
-}
 
-- (void)testDeleteDocument {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
-  [self writeDocumentRef:doc data:@{@"foo" : @"bar"}];
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
-  XCTAssertEqualObjects(@"bar", snapshot[@"foo"]);
+  [_testCase awaitExpectations];
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
+- (void)runFailingTransaction:(FIRFirestoreErrorCode)expected {
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"runTransaction"];
+  [_db
       runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        [transaction deleteDocument:doc];
+        for (TransactionStage stage in self->_stages) {
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertEqualObjects(@YES, result);
-        XCTAssertNil(error);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence (%@), to fail, but it didn't.",
+                                       [self listStages]];
+        [self->_testCase assertErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
 
-  snapshot = [self readDocumentForRef:doc];
+  [_testCase awaitExpectations];
+}
+
+- (void)cleanupTester {
+  _stages = [NSArray array];
+  // Set the docRef to something else to lose the original reference.
+  _docRef = [[self->_db collectionWithPath:@"reset"] documentWithAutoID];
+}
+
+- (NSString *)listStages {
+  NSMutableArray<NSString *> *seqList = [NSMutableArray array];
+  for (TransactionStage stage in _stages) {
+    if (stage == delete1) {
+      [seqList addObject:@"delete"];
+    } else if (stage == update1 || stage == update2) {
+      [seqList addObject:@"update"];
+    } else if (stage == set1 || stage == set2) {
+      [seqList addObject:@"set"];
+    } else if (stage == get) {
+      [seqList addObject:@"get"];
+    }
+  }
+  return [seqList description];
+}
+
+@end
+
+@implementation FSTTransactionTests
+- (void)assertExistsWithSnapshot:(FIRDocumentSnapshot *)snapshot error:(NSError *)error {
```

Comment:
It's probably worth separating these from the tests proper so that you can comment on why you need them.

I suggest adding category for these assertions. Concretely this means adding this above:

```objc
@interface FSTTransactionTests (Assertions)
// add all your assertion declarations
@end
```

... and then moving these assertion helper definitions into a separate

```objc
@implementation FSTTransactionTests (Assertions)
// assertion definitions go here.
@end
```

Then add a comment block to the category declaration that describes why you're doing what you're doing (namely that XCTest assertions don't work in classes that don't extend XCTestCase).

Alternatively, if you don't like the category idea as a means of grouping these separately, at least add the same block comment to the main declaration.

Reply:
woah, categories are a pretty cool concept!